src/Microsoft.AspNetCore.Razor.Evolution/TagHelperBinderSyntaxTreePass.cs

```diff
@@ -22,18 +24,33 @@ public RazorSyntaxTree Execute(RazorCodeDocument document, RazorSyntaxTree synta
                 return syntaxTree;
             }
 
-            var errorSink = new ErrorSink();
-            var visitor = new TagHelperDirectiveSpanVisitor(resolver, errorSink);
-            var descriptors = visitor.GetDescriptors(syntaxTree.Root);
+            // We need to find errors in all of the *imports* as well as in the main razor file
+            //
+            // The imports come logically before the main razor file and are in the order they
+            // should be processed.
+            var visitor = new Visitor();
+            var imports = codeDocument.GetImportSyntaxTrees();
+            if (imports != null)
+            {
+                for (var i = 0; i < imports.Count; i++)
+                {
+                    var import = imports[i];
+                    visitor.VisitBlock(import.Root);
+                }
+            }
+            
+            visitor.VisitBlock(syntaxTree.Root);
 
-            if (!descriptors.Any())
+            var directives = visitor.Directives;
+            var errorSink = new ErrorSink();
+            var descriptors = resolver.Resolve(new TagHelperDescriptorResolutionContext(directives, errorSink)).ToArray();
+            if (descriptors.Length == 0)
             {
+                // No TagHelpers, add any errors if we have them.
                 if (errorSink.Errors.Count > 0)
                 {
-                    var combinedErrors = CombineErrors(syntaxTree.Diagnostics, errorSink.Errors);
-                    var erroredTree = RazorSyntaxTree.Create(syntaxTree.Root, syntaxTree.Source, combinedErrors, syntaxTree.Options);
-
-                    return erroredTree;
+                    var errors = CombineErrors(syntaxTree.Diagnostics, errorSink.Errors);
+                    return RazorSyntaxTree.Create(syntaxTree.Root, syntaxTree.Source, errors, syntaxTree.Options);
```

Comment:
Unfortunately I can't add test for this, we don't actually process the tag helper directives today, that's next.